File: 0234-palindrome-linked-list/0234-palindrome-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
   ListNode*reverselist(ListNode*head){
      ListNode*prev=NULL;
      ListNode*curr=head;
      ListNode*frwd=curr;
      while(curr!=NULL){
        frwd=curr->next;
        curr->next=prev;
        prev=curr;
        curr=frwd;
      }
      return prev;


   }
    bool isPalindrome(ListNode* head) {

       if(head->next==NULL){
        return true;
       }
       ListNode*slow=head;
       ListNode*fast=head;
       while(fast!=NULL && fast->next!=NULL){
        fast=fast->next->next;
        slow=slow->next;
       }
       ListNode*right=reverselist(slow);
       while(right!=NULL){
        if(right->val!=head->val)
         return false;
        head=head->next;
        right=right->next;
       }
       return true;




        // ListNode*ptr=head;
        // if(head->next==NULL){
        //     return true;
        // }
        // vector<int>ans;
        // while(ptr!=NULL){
        //     ans.push_back(ptr->val);
        //     ptr=ptr->next;
        // }
        // vector<int>res=ans;
        // reverse(res.begin(),res.end());
        // if(ans==res)
        //     return true;
        // return false;
        
    }
};

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverse(ListNode slow){
        ListNode frwd=slow;
        ListNode prev=null;
        ListNode curr=slow;
        while(frwd!=null){
            frwd=frwd.next;
            curr.next=prev;
            prev=curr;
            curr=frwd;
        }
        return prev;
    }

    public boolean isPalindrome(ListNode head) {
        ListNode slow=head;
        ListNode fast=head;
        if(head==null || head.next==null)return true;
        while(fast!=null && fast.next!=null){
            slow=slow.next;
            fast=fast.next.next;
        }
        slow=reverse(slow);
        while(slow!=null){
            if(slow.val!=head.val)return false;
            slow=slow.next;
            head=head.next;
        }
        return true;
    }
}